R/zchunk_LB171.LC_R_Cropland_Yh_GLU_irr.R

```diff
@@ -51,82 +51,48 @@ module_aglu_LB171.LC_R_Cropland_Yh_GLU_irr <- function(command, ...) {
-
-    L122.LC_bm2_R_HarvCropLand_C_Yh_GLU %>%
-      # Total cropland cover, will be used for where rainfed share is missing
-      rename(total = value) %>%
-      # Match the base year share by GCAM region, commodity, year and GLU.
-      left_join_error_no_match(L171.ag_irrHA_frac_R_C_GLU, by = c("GCAM_region_ID", "GCAM_commodity", "GLU")) %>%
-      # Calculate the rainfed cropland cover in all historical periods.
-      mutate(value = total * rfd_share,
-             # For the rainfed cropland table, missing values default to total cropland quantities; there are no missing currently (5/10/17).
-             value = if_else(is.na(value), total, value)) %>%
-      select(GCAM_region_ID, GCAM_commodity, GLU, year, value) ->
-      L171.LC_bm2_R_rfdHarvCropLand_C_Yh_GLU
-
-    # Extend irrigated cropland cover data to all years, 1700-2010
-    # Note: 1700, 1750, 1800, 1850, 1900, and 1950 are all zero; and this interpolation fills 1700-1970 with zeros, but no missing elsewhere.
-    L171.LC_bm2_R_irrHarvCropLand_C_Yh_GLU %>%
-      select(GCAM_region_ID, GCAM_commodity, GLU) %>%
-      unique() -> id_list
-    tibble(year = seq(1700, max(HISTORICAL_YEARS))) -> all_years
-    id_list %>%
-      repeat_add_columns(all_years) %>%
-      left_join(L171.LC_bm2_R_irrHarvCropLand_C_Yh_GLU, by = c("GCAM_region_ID", "GCAM_commodity", "GLU", "year")) %>%
-      group_by(GCAM_region_ID, GCAM_commodity, GLU) %>%
-      mutate(value = approx_fun(year, value)) %>%
-      ungroup() -> irrHarvCropLand_interp
-
-    # Extend rainfed and cropland cover data to all years, 1700-2010
-    # Note: 1700, 1750, 1800, 1850, 1900, and 1950 are all zero; and this interpolation fills 1700-1970 with zeros, but no missing elsewhere.
-    L171.LC_bm2_R_rfdHarvCropLand_C_Yh_GLU %>%
-      select(GCAM_region_ID, GCAM_commodity, GLU) %>%
-      unique() -> id_list
-    tibble(year = seq(1700, max(HISTORICAL_YEARS))) -> all_years
-    id_list %>%
-      repeat_add_columns(all_years) %>%
-      left_join(L171.LC_bm2_R_rfdHarvCropLand_C_Yh_GLU, by = c("GCAM_region_ID", "GCAM_commodity", "GLU", "year")) %>%
+    IrrRfdCropland <-
+        L122.LC_bm2_R_HarvCropLand_C_Yh_GLU %>%
+          left_join_error_no_match(L171.ag_irrHA_frac_R_C_GLU,
+                                   by = c("GCAM_region_ID", "GCAM_commodity",
+                                   "GLU")) %>%
+          mutate(irr.harvarea = value * irrHA_frac,
+                 irr.harvarea = replace(irr.harvarea, is.na(irr.harvarea), 0.0), # if no data, assume irrigated is zero
+                 rfd.harvarea = value * rfd_share,
+                 rfd.harvarea = replace(rfd.harvarea, is.na(rfd.harvarea), value)) # if no data, assume all rainfed
+
+    ## Extend to cover all years 1700-2010.
+    idvars <- c('GCAM_region_ID', 'GCAM_commodity', 'GLU', 'year')
+    allyr <- seq(min(HISTORICAL_YEARS), max(HISTORICAL_YEARS))
+    IrrRfdCropland %>%
+      tidyr::expand(nesting(GCAM_region_ID, GCAM_commodity, GLU),
```

Comment:
That's correct.  Basically (and I kind of had to piece this together because the docs for tidyr are a little sparse), `expand` does two things:
1. Interprets names in the context of the table in its first argument (`IrrRfdCropland` in this case).  As far as I can tell, that's the only thing the first argument is used for.
2. Returns as a tbl_df the cross product of its remaining arguments.  Using named arguments allows you to control the names of the columns in the resulting table.  

In this case it has two arguments.  One is `allyr`, which we name as `year`.  The other is `nesting(GCAM_region_ID, GCAM_commodity, GLU)`, which returns the combinations of those three variables that actually occur in the data set.  (As far as I can make out, this is equivalent to `cbind(GCAM_region_ID, GCAM_commodity, GLU)`, except for the class of the result.

This gives us a table with columns `GCAM_region_ID`, `GCAM_commodity`, `GLU`, and `year`.  When we join this to the original data frame (`IrrRfdCropland`), we get back the remaining columns from the original data, with `NA` filled in for combinations that didn't exist before.  I _think_ that this last step is what `complete` is supposed to do, but I've never been able to get it to work right for some reason, so I find it easier just to write it out explicitly.